src/D3MHub.sol

```diff
@@ -0,0 +1,601 @@
+        @param what bytes32("tau") or it will revert
+        @param data number of seconds to wait after caging a pool to write off debt
+    */
+    function file(bytes32 ilk, bytes32 what, uint256 data) external auth {
+        if (what == "tau") ilks[ilk].tau = data;
+        else revert("D3MHub/file-unrecognized-param");
+
+        emit File(ilk, what, data);
+    }
+
+    /**
+        @notice update plan or pool addresses for D3M ilk.
+        @dev msg.sender must be authorized.
+        @param ilk  bytes32 of the D3M ilk to be updated
+        @param what bytes32("pool"|"plan") or it will revert
+        @param data address we are setting it to
+    */
+    function file(bytes32 ilk, bytes32 what, address data) external auth {
+        require(vat.live() == 1, "D3MHub/no-file-during-shutdown");
+        require(ilks[ilk].tic == 0, "D3MHub/pool-not-live");
+
+        if (what == "pool") ilks[ilk].pool = ID3MPool(data);
+        else if (what == "plan") ilks[ilk].plan = ID3MPlan(data);
+        else revert("D3MHub/file-unrecognized-param");
+        emit File(ilk, what, data);
+    }
+
+    // --- Deposit controls ---
+    function _wind(bytes32 ilk, ID3MPool _pool, uint256 amount) internal {
+        // IMPORTANT: this function assumes Vat rate of D3M ilks will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+        if (amount == 0) {
+            emit Wind(ilk, 0);
+            return;
+        }
+
+        vat.slip(ilk, address(_pool), int256(amount));
+        vat.frob(ilk, address(_pool), address(_pool), address(this), int256(amount), int256(amount));
+        // normalized debt == erc20 DAI (Vat rate for D3M ilks fixed to 1 RAY)
+        daiJoin.exit(address(_pool), amount);
+        _pool.deposit(amount);
+
+        emit Wind(ilk, amount);
+    }
+
+    function _unwind(
+                bytes32 ilk,
+                ID3MPool _pool,
+                uint256 supplyReduction, // [wad]
+                Mode mode,
+                uint256 assetBalance     // [wad]
+             ) internal {
+        // IMPORTANT: this function assumes Vat rate of D3M ilks will always be == 1 * RAY (no fees).
+        // That's why it converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+        // This module will have an unintended behaviour if rate is changed to some other value.
+
+        EndLike _end;
+        uint256 daiDebt;
+        if (mode == Mode.NORMAL) {
+            // Normal mode or module just caged (no culled)
+            // debt is obtained from CDP art
+            (,daiDebt) = vat.urns(ilk, address(_pool));
+        } else if (mode == Mode.MODULE_CULLED) {
+            // Module shutdown and culled
+            // debt is obtained from free collateral owned by this contract
```

Comment:
free collateral is owned by the pool, not "this contract"

Reply:
0b42cfa